cmd/ipfs/daemon.go

```diff
@@ -176,9 +194,59 @@ func listenAndServeAPI(node *core.IpfsNode, req cmds.Request, addr ma.Multiaddr)
 	case <-node.Closing():
 		log.Infof("daemon at %s terminating...", addr)
 		server.Shutdown <- true
-		<-serverExited // now, DO wait until server exits
+		<-serverExited // now, DO wait until server exit
 	}
 
 	log.Infof("daemon at %s terminated", addr)
 	return serverError
 }
+
+// the gateway also listens on its own address:port in addition to the API listener
+func listenAndServeGateway(node *core.IpfsNode, addr ma.Multiaddr) error {
+	_, host, err := manet.DialArgs(addr)
+	if err != nil {
+		return err
+	}
+
+	server := manners.NewServer()
+	mux := http.NewServeMux()
+	gateway, err := NewGatewayHandler(node)
+	if err != nil {
+		return err
+	}
+	mux.Handle("/ipfs/", gateway)
+
+	done := make(chan struct{}, 1)
+	defer func() {
+		done <- struct{}{}
+	}()
```

Comment:
this sync is incorrect I think. this will be called right after exiting this function, which will drop the other goroutine. what is this done channel supposed to do?